forte/data/data_pack_dataset.py

```diff
@@ -0,0 +1,278 @@
+#  Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, Iterator, Type, Optional, List, Tuple, Union, Any
+
+import torch
+from texar.torch import HParams
+from texar.torch.data import IterDataSource, DatasetBase, Batch
+
+from forte.data.converter import Converter
+from forte.data.converter import Feature
+from forte.data.data_pack import DataPack
+from forte.data.extractor.base_extractor import BaseExtractor
+from forte.data.ontology.core import EntryType
+from forte.data.ontology.top import Annotation
+from forte.data.types import DataRequest
+
+__all__ = [
+    "DataPackIterator",
+    "DataPackDataSource",
+    "DataPackDataset"
+]
+
+# An instance is a single data point from data pack
+Instance = Dict
+RawExample = Tuple[Instance, DataPack]
+FeatureCollection = Dict[str, Feature]
+
+
+class DataPackIterator:
+    """
+    An iterator over single data example from multiple data packs.
+
+    Args:
+        pack_generator (Iterator[DataPack]): A generator of
+            :class:`forte.data.data_pack.DataPack`.
+        context_type: The granularity of a single example which
+            could be any ``Annotation`` type. For example, it can be
+            :class:`ft.onto.base_ontology.Sentence`, then each training example
+            will represent the information of a sentence.
+        request: The request of type `Dict` sent to
+            :class:`forte.data.readers.base_reader.PackReader` to query
+            specific data.
+        skip_k (int): Will skip the first `skip_k` instances and generate
+            data from the (`offset` + 1)th instance.
+
+    .. note::
```

Comment:
The iterator return value seems to be missing. I think you mean to return a tuple of instance and data pack.

Also, it will be helpful to provide an example usage of this iterator seems it is slightly complex.

Reply:
Fixed and an example is added.